File: built-in plugins dsp/mb-woodwind-bassoon.cpp
/**
 * MB Bassoon
 * Category : instrument
 * Type     : woodwind
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Deep bassoon with dark reedy character
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WOODWIND_BASSOON_H
#define MB_WOODWIND_BASSOON_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWoodwindBassoon : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-woodwind-bassoon";
    static constexpr const char* PLUGIN_NAME    = "MB Bassoon";
    static constexpr const char* PLUGIN_TYPE    = "woodwind";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float reed = 0.5f;  // range [0, 1]
    float body = 0.7f;  // range [0, 1]
    float vibrato = 0.2f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbWoodwindBassoon() = default;
    ~MbWoodwindBassoon() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.reed = std::clamp(params.reed, 0f, 1f);
        params.body = std::clamp(params.body, 0f, 1f);
        params.vibrato = std::clamp(params.vibrato, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Bassoon
        return input;
    }
};

#endif // MB_WOODWIND_BASSOON_H

File: built-in plugins dsp/mb-woodwind-bassoon.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbWoodwindBassoonPlugin: PluginDefinition = {
  id: "mb-woodwind-bassoon",
  slug: "mb-woodwind-bassoon",
  name: "MB Bassoon",
  category: "instrument",
  type: "woodwind" as any,
  version: "1.0.0",
  description: "Deep bassoon with dark reedy character",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.45 },
    { type: "square", detune: 0, gain: 0.3 },
    { type: "sine", detune: -1200, gain: 0.25 },
  ],
  envelope: { attack: 0.06, decay: 0.3, sustain: 0.82, release: 0.18 },
  parameters: [
    {
      id: "reed",
      name: "Reed",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "body",
      name: "Body",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "vibrato",
      name: "Vibrato",
      type: "float",
      defaultValue: 0.2,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { reed: 0.5, body: 0.7, vibrato: 0.2, volume: 0.8 },
};

export default MbWoodwindBassoonPlugin;
